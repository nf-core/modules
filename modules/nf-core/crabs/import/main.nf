process CRABS_IMPORT {
    tag "$meta.id"
    label 'process_medium'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/crabs:1.0.7--pyhdfd78af_0':
        'biocontainers/crabs:1.0.7--pyhdfd78af_0' }"

    input:
    tuple val(meta), path(fasta)
    tuple val(meta2), path(accession2taxid)
    tuple val(meta3), path(names)
    tuple val(meta4), path(nodes)
    val(import_format)

    output:
    tuple val(meta), path("*.txt"), emit: crabsdb
    path "versions.yml"           , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args           = task.ext.args ?: ''
    def prefix         = task.ext.prefix ?: "${meta.id}"
    def is_compressed  = fasta.name.endsWith(".gz")
    def fasta_name     = fasta.name.replace(".gz", "")
    def import_fmt_cmd = "--import-format ${import_format}"
    def version_cmd    = "\$(crabs --help 2>/dev/null | grep 'CRABS |' | sed 's/.*CRABS | v\\([0-9.]*\\).*/\\1/')"
    if (workflow.profile.tokenize(',').intersect(['conda', 'mamba']).size() >= 1) {
        version_cmd    = '1.0.7'
    }
    """
    if [ "${is_compressed}" == "true" ]; then
        gzip -c -d ${fasta} > ${fasta_name}
    fi

    crabs --import \\
        --input ${fasta_name} \\
        --output ${prefix}.txt \\
        --acc2tax ${accession2taxid} \\
        --names ${names} \\
        --nodes ${nodes} \\
        ${import_fmt_cmd} \\
        ${args}

    rm ${fasta_name}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        crabs: ${version_cmd}
    END_VERSIONS
    """

    stub:
    def prefix       = task.ext.prefix ?: "${meta.id}"
    def version_cmd  = "\$(crabs --help 2>/dev/null | grep 'CRABS |' | sed 's/.*CRABS | v\\([0-9.]*\\).*/\\1/')"
    if (workflow.profile.tokenize(',').intersect(['conda', 'mamba']).size() >= 1) {
        version_cmd  = '1.0.7'
    }
    """
    touch ${prefix}.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        crabs: ${version_cmd}
    END_VERSIONS
    """
}
