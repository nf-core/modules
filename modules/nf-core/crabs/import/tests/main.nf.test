nextflow_process {

    name "Test Process CRABS_IMPORT"
    script "../main.nf"
    process "CRABS_IMPORT"

    tag "modules"
    tag "modules_nfcore"
    tag "crabs"
    tag "crabs/import"

    test("sarscov2 - fasta") {

        config "./nextflow.config"

        when {
            params {
                module_args = "--ranks 'superkingdom;phylum;class;order;family;genus;species'"
            }
            process {
                """
                input[0] = [
                    [ id:'test' ], // meta map
                    file(params.modules_testdata_base_path +'genomics/sarscov2/genome/genome-ena.fasta', checkIfExists: true)
                ]
                input[1] = [
                    [ id:'accession2taxid' ], // meta map
                    file(params.modules_testdata_base_path +'genomics/prokaryotes/metagenome/taxonomy/accession2taxid/nucl_gb.accession2taxid', checkIfExists: true)
                ]
                input[2] = [
                    [ id:'names' ], // meta map
                    file(params.modules_testdata_base_path +'genomics/prokaryotes/metagenome/taxonomy/taxdmp/names.dmp', checkIfExists: true)
                ]
                input[3] = [
                    [ id:'nodes' ], // meta map
                    file(params.modules_testdata_base_path +'genomics/prokaryotes/metagenome/taxonomy/taxdmp/nodes.dmp', checkIfExists: true)
                ]
                input[4] = 'embl' // import format
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("sarscov2 - fasta - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = [
                    [ id:'test' ], // meta map
                    file(params.modules_testdata_base_path +'genomics/sarscov2/genome/genome-ena.fasta', checkIfExists: true)
                ]
                input[1] = [
                    [ id:'accession2taxid' ], // meta map
                    file(params.modules_testdata_base_path +'genomics/prokaryotes/metagenome/taxonomy/accession2taxid/nucl_gb.accession2taxid', checkIfExists: true)
                ]
                input[2] = [
                    [ id:'names' ], // meta map
                    file(params.modules_testdata_base_path +'genomics/prokaryotes/metagenome/taxonomy/taxdmp/names.dmp', checkIfExists: true)
                ]
                input[3] = [
                    [ id:'nodes' ], // meta map
                    file(params.modules_testdata_base_path +'genomics/prokaryotes/metagenome/taxonomy/taxdmp/nodes.dmp', checkIfExists: true)
                ]
                input[4] = 'embl' // import format
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out,
                    path(process.out.versions[0]).yaml
                    ).match() }
            )
        }

    }

}