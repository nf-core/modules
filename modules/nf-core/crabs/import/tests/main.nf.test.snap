{
    "sarscov2 - fasta - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,d21925bd41f67039425ac09278c847d0"
                ],
                "crabsdb": [
                    [
                        {
                            "id": "test"
                        },
                        "test.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,d21925bd41f67039425ac09278c847d0"
                ]
            },
            {
                "CRABS_IMPORT": {
                    "crabs": "1.0.7"
                }
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-06-02T10:22:19.14791"
    },
    "sarscov2 - fasta": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.txt:md5,22c8c613e67ce3f5237a6c4646a0abb1"
                    ]
                ],
                "1": [
                    "versions.yml:md5,d21925bd41f67039425ac09278c847d0"
                ],
                "crabsdb": [
                    [
                        {
                            "id": "test"
                        },
                        "test.txt:md5,22c8c613e67ce3f5237a6c4646a0abb1"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,d21925bd41f67039425ac09278c847d0"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-06-02T12:55:24.000777"
    }
}